### PR TITLE
making WarpDrive more generic

### PR DESCRIPTION
**Pull request by sunil-s**
- adapt function manager with higher-dimensional (>=3) features;
-  add gym-style step() and reset() APIs to the env wrapper

### CODE REVIEW

**Review by Emerald01 — APPROVED** [review]

File: warp_drive/utils/env_registrar.py [hunk 1]
```diff
@@ -5,23 +5,27 @@ class CustomizedEnvironmentRegistrar:
...
-                f"but we will re-register it by overwriting the previous source code path "
+                f"[WARNING]: EnvironmentRegistrar has already registered an "
+                f"environment called {env_name} but we will re-register it "
+                f"by overwriting the previous source code path"
             )
-        assert "cu" == cuda_env_src_path.rsplit(".", 1)[1], \
-            "the customzed environment is expected to be a CUDA source code (*.cu)"
+        assert (
+            "cu" == cuda_env_src_path.rsplit(".", 1)[1]
+        ), "the customzed environment is expected to be a CUDA source code (*.cu)"
```
Comment (StephanZheng) [inline warp_drive/utils/env_registrar.py:23]: typo: customized or custom?

File: warp_drive/utils/unittest_run.py [hunk 2]
```diff
@@ -19,4 +20,4 @@
     except subprocess.TimeoutExpired:
         test_process.kill()
         outs, errs = test_process.communicate()
-        print("Unittest Timeout")
\ No newline at end of file
+        print("Unittest Timeout")
```
Comment (StephanZheng) [inline warp_drive/utils/unittest_run.py:23]: Unit Test Timeout

**Review by StephanZheng — CHANGES_REQUESTED** [review]
typos -- will check logic later

File: warp_drive/training/trainer.py [hunk 3]
```diff
@@ -249,6 +219,88 @@ def generate_rollout(self, batch_index):
...
+        for action_idx, probs in enumerate(probabilities):
+            action_name = f"{_ACTIONS}_{action_idx}" + suffix
+            self.cuda_sample_controller.sample(
+                self.cuda_envs.cuda_data_manager, probs, action_name
+            )
+            actions = self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
+                action_name
+            )
+
+            self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:278]: If there is one single policy + one discrete action space, then there is no need to create a mapping buffer and copy it back. This will save time for example, shared policy for tag gridworld

File: warp_drive/training/trainer.py [hunk 4]
```diff
@@ -40,6 +41,8 @@ def __init__(
         env_wrapper=None,
         config=None,
         policy_tag_to_agent_id_map=None,
+        create_separate_placeholders_for_each_policy=False,
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:44]: I think it is necessary to add docstring to explain those arguments

File: warp_drive/training/trainer.py [hunk 5]
```diff
@@ -249,6 +219,88 @@ def generate_rollout(self, batch_index):
...
         self.perf_stats.env_step_time += start_event.elapsed_time(end_event) / 1000
 
+    def evaluate_policies(self, batch_index):
+        probabilities = {}
+        for policy in self.policies:
+            probabilities[policy], _ = self.models[policy](
+                batch_index=batch_index, batch_size=self.training_batch_size_per_env
+            )
+            # Combine probabilities if there are multiple policies,
+            # yet they share the same placeholders
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:229]: the final probabilities should concatenate in terms of the agent dim (i.e., policies) since we do not have policy placeholder in this case, like [[num_envs, num_agents, action_dim],[...]]

File: warp_drive/training/trainer.py [hunk 6]
```diff
@@ -249,6 +227,105 @@ def generate_rollout(self, batch_index):
...
+            )
+
+        # Combine probabilities across policies if there are multiple policies,
+        # yet they are expected to share the same action placeholders
+        if (
+            len(self.policies) > 1
+            and not self.create_separate_placeholders_for_each_policy
+        ):
+            for policy in self.policies:
+                num_action_types[policy] = len(probabilities[policy])
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:244]: `num_action_types` seems not have a variable declaration

File: warp_drive/training/trainer.py [hunk 7]
```diff
@@ -249,6 +227,105 @@ def generate_rollout(self, batch_index):
...
+                probability_shapes[policy] = [
+                    probabilities[policy][action_idx].shape
+                    for action_idx in range(num_action_types[policy])
+                ]
+            assert all_equal(list(num_action_types.values()))
+
+            # Initialize combined_probabilities
+            first_policy = list(probabilities.keys())[0]
+            num_action_types = len(probabilities[first_policy])
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:253]: `num_action_types` is already used above

File: warp_drive/training/trainer.py [hunk 8]
```diff
@@ -249,6 +227,105 @@ def generate_rollout(self, batch_index):
...
+
+            first_action_idx = 0
+            num_envs = probabilities[first_policy][first_action_idx].shape[0]
+            action_dim = probabilities[first_policy][first_action_idx].shape[-1]
```
Comment (Emerald01) [inline warp_drive/training/trainer.py:257]: `action_dim` is dependent on `action_type`, may not be defined outside of the loop as a constant, right?

Comment (sunil-s) [reply warp_drive/training/trainer.py:257]: yes, correct.

Comment (Emerald01) [reply warp_drive/training/trainer.py:288]: This line can be removed

File: warp_drive/training/models/fully_connected.py [hunk 9]
```diff
@@ -69,53 +98,139 @@ def __init__(self, env, model_config, policy, policy_tag_to_agent_id_map):
...
-                self.env.cuda_data_manager.data_on_device_via_torch(name=name)[
-                    batch_index, :, agent_ids_for_policy
-                ] = obs[:, agent_ids_for_policy]
                 ip = obs[:, agent_ids_for_policy]
             else:
-                self.env.cuda_data_manager.data_on_device_via_torch(name=name)[
-                    batch_index
-                ] = obs
                 ip = obs
+
```
Comment (Emerald01) [inline warp_drive/training/models/fully_connected.py:203]: This is optional though. I feel this forward function serves two purposes, one is forward prediction, the other is feed obs batch, is it better to factorize a function to do the 2nd job here because it might be confusing people

Comment (sunil-s) [reply warp_drive/training/models/fully_connected.py:203]: Yes, I think it makes a lot of sense.

File: warp_drive/training/trainer.py [hunk 10]
```diff
@@ -235,20 +229,155 @@ def generate_rollout(self, batch_index):
...
+        if num_action_types == 1:
+            action_name = _ACTIONS + suffix
+            self.cuda_sample_controller.sample(
+                self.cuda_envs.cuda_data_manager, probabilities[0], action_name
+            )
+            actions = self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
+                action_name
+            )
+            self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
+                name=f"{_ACTIONS}_batch"
```
Comment (sunil-s) [inline warp_drive/training/trainer.py:337]: I believe this should be name=f"{_ACTIONS}_batch" + suffix
